File: backend/app/schemas/prompt_eval.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, model_validator


class PromptEvalCriterionInfo(BaseModel):
    id: str
    label: str
    category: str
    method: Literal["deterministic", "judged"]
    weight: int
    description: str


class PromptEvalRequest(BaseModel):
    agent_id: uuid.UUID | None = None
    prompt_text: str | None = None
    # "effective" (base_instruction + attached skills, composed exactly as
    # the runtime sends it) is the more useful default for an agent-scoped
    # evaluation — it's what the model actually receives. Ignored (always
    # treated as "static") when evaluating raw pasted text, since there are
    # no skills to compose against.
    scope: Literal["static", "effective"] = "effective"
    model: str | None = None

    @model_validator(mode="after")
    def _require_exactly_one_input(self) -> "PromptEvalRequest":
        has_agent = self.agent_id is not None
        has_text = bool(self.prompt_text and self.prompt_text.strip())
        if not has_agent and not has_text:
            raise ValueError("Provide either agent_id or prompt_text.")
        if has_agent and has_text:
            raise ValueError("Provide only one of agent_id or prompt_text, not both.")
        return self


class CriterionResultOut(BaseModel):
    id: str
    label: str
    category: str
    method: Literal["deterministic", "judged"]
    weight: int
    score: int | None
    max_score: int
    applicable: bool
    severity: Literal["info", "warning", "critical"]
    rationale: str
    suggestion: str | None = None


class PromptEvalResult(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID | None
    agent_name: str | None
    scope: str
    source_text: str
    overall_score: float
    criteria: list[CriterionResultOut]
    summary: str | None
    suggested_rewrite: str | None
    model_used: str | None
    judge_error: str | None
    created_at: datetime


class PromptEvalRunSummary(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID | None
    agent_name: str | None
    scope: str
    overall_score: float
    summary: str | None
    model_used: str | None
    judge_error: str | None
    created_by: str | None
    created_at: datetime

File: frontend/src/api/promptEval.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { api } from './client'

export type PromptEvalScope = 'static' | 'effective'
export type PromptEvalMethod = 'deterministic' | 'judged'
export type PromptEvalSeverity = 'info' | 'warning' | 'critical'

export interface PromptEvalCriterionInfo {
  id: string
  label: string
  category: string
  method: PromptEvalMethod
  weight: number
  description: string
}

export interface PromptEvalRequest {
  agent_id?: string
  prompt_text?: string
  scope?: PromptEvalScope
  model?: string
}

export interface CriterionResultOut {
  id: string
  label: string
  category: string
  method: PromptEvalMethod
  weight: number
  score: number | null
  max_score: number
  applicable: boolean
  severity: PromptEvalSeverity
  rationale: string
  suggestion: string | null
}

export interface PromptEvalResult {
  id: string
  agent_id: string | null
  agent_name: string | null
  scope: string
  source_text: string
  overall_score: number
  criteria: CriterionResultOut[]
  summary: string | null
  suggested_rewrite: string | null
  model_used: string | null
  judge_error: string | null
  created_at: string
}

export interface PromptEvalRunSummary {
  id: string
  agent_id: string | null
  agent_name: string | null
  scope: string
  overall_score: number
  summary: string | null
  model_used: string | null
  judge_error: string | null
  created_by: string | null
  created_at: string
}

const CRITERIA_KEY = ['prompt-eval', 'criteria'] as const
const RUNS_KEY = ['prompt-eval', 'runs'] as const

export function usePromptEvalCriteria() {
  return useQuery({
    queryKey: CRITERIA_KEY,
    queryFn: () => api.get<PromptEvalCriterionInfo[]>('/prompt-eval/criteria'),
    staleTime: Infinity, // static catalog — never changes without a deploy
  })
}

export function useEvaluatePrompt() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (payload: PromptEvalRequest) => api.post<PromptEvalResult>('/prompt-eval/evaluate', payload),
    onSuccess: () => qc.invalidateQueries({ queryKey: RUNS_KEY }),
  })
}

export function usePromptEvalRuns(agentId?: string) {
  return useQuery({
    queryKey: [...RUNS_KEY, agentId ?? 'all'],
    queryFn: () =>
      api.get<PromptEvalRunSummary[]>(`/prompt-eval/runs${agentId ? `?agent_id=${agentId}` : ''}`),
  })
}

export function usePromptEvalRun(id: string | undefined) {
  return useQuery({
    queryKey: [...RUNS_KEY, 'detail', id ?? ''],
    queryFn: () => api.get<PromptEvalResult>(`/prompt-eval/runs/${id}`),
    enabled: !!id,
  })
}
